### PR TITLE
[REF] Add a column containing an expiry_date to civicrm_prevnext_cache tabl…

### PR DESCRIPTION
**Pull request by seamuslee001**
…e, remove prevnext_cache group as no longer needed and add a flush method to prevNextCache Interface and remove all prevnext records that have expired when calling flush

Overview
----------------------------------------
This does a number of things

1. adds in an expiry_date column to civicrm_prevnext_cache
2. adds in a flush method to the `Civi::service('prevnext')` interface to clear out the cache only really mattering for when it is sqlbackend not so much redis as redis auto expires cache
3. removes the Civi::cache('prevnext') cache group as no longer required in this model
4. Updates various functions that insert or update data in civicrm_prevnext_cache to set an appropriate expiry date based on a default TTL

Before
----------------------------------------
PrevNext Service and PrevNExt Cache group required in SQL Backed instances

After
----------------------------------------
Only PrevNext Service required for both SQL and Redis backed instances

ping @eileenmcnaughton @totten

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@seamuslee001 can you rebase?

**Comment by eileenmcnaughton** [comment]
@seamuslee001 my first thought about this was that it makes a lot of sense but that we should follow the pattern for the civicrm_cache table - ie https://github.com/civicrm/civicrm-core/compare/master...eileenmcnaughton:expired?expand=1

But when I looked at that I had doubts - we are going from setting the expiry at the key level to at the row (item within the key level) - with the granularity of a timestamp field we need to be really careful the rows don't save with slightly different expiry dates (probably you have dealt with that but it feels like a risk if we switch to tracking expiry by row).

I'd want to hear @totten's thoughts on this change of approach

**Comment by seamuslee001** [comment]
@eileenmcnaughton it feels like we are on a row based caching in redis implementation currently but I maybe not fully understanding the redis storage.

File: CRM/Core/Payment/PayPalIPN.php [hunk 1]
```diff
@@ -398,9 +398,7 @@ public function getInput(&$input, &$ids) {
        * The `payment_date` that Paypal sends back is in their timezone. Example return: 08:23:05 Jan 11, 2019 PST
        * Subsequently, we need to account for that, otherwise the recieve time will be incorrect for the local system
        */
-      $systemTimeZone = new DateTimeZone(CRM_Core_Config::singleton()->userSystem->getTimeZoneString());
-      $receiveDateTime->setTimezone($systemTimeZone);
-      $input['receive_date'] = $receiveDateTime->format('YmdHis');
+      $input['receive_date'] = CRM_Utils_Date::convertDateToLocalTime($receiveDateTime);
```
Comment (totten) [inline CRM/Core/Payment/PayPalIPN.php]: (1) I like how the name of the extracted function adds more clarity/pithiness.

(2) Small nitpick: I did a double-take wondering how `PayPalIPN.php` is related to the PR topic. Since the method-extraction is a bit tangential to the overall theme of adding `civicrm_prevnext_cache.expiry_date`, it might've made sense to do it as a separate commit. (Not critical, though.)

**Comment by eileenmcnaughton** [comment]
I would also note that civicrm_cache uses expired_date not expiry_date - if we go this way then I feel they should be the same

File: CRM/Core/PrevNextCache/Interface.php [hunk 2]
```diff
@@ -130,4 +130,9 @@ public function getCount($cacheKey);
    */
   public function fetch($cacheKey, $offset, $rowCount);
 
+  /**
+   * Remove expired cache objects.
+   */
+  public function flush();
```
Comment (totten) [inline CRM/Core/PrevNextCache/Interface.php]: In `CRM_Utils_Cache_Interface`, the `flush()` operation is one which proactively removes *all* items from the named cache. I generally perceive "flush" as an indiscriminate action.

It sounds like this operation is meant more as an incremental cleanup/timeout for expired items. For something that is incremental/targetted, I think `prune()` or `cleanup()` might be more evocative.

File: xml/schema/Core/PrevNextCache.xml [hunk 3]
```diff
@@ -72,4 +72,17 @@
     <fieldName>entity_table</fieldName>
     <fieldName>is_selected</fieldName>
   </index>
+  <field>
+    <name>expiry_date</name>
+    <title>Prev Next Expiry Date</title>
+    <type>timestamp</type>
+    <default>current_timestamp()</default>
```
Comment (totten) [inline xml/schema/Core/PrevNextCache.xml]: :+1: for `timestamp`.

The name `expiry_date` stuck out as me as unfamiliar and had me thinking, "we should be consistent!" Well, it turns out... this is consistent! Or, more strictly, every field involving the end times uses a different name (`expired_date`, `expiry_date`, `expires_on`,  `expiration_time`, `hush_until`). One of them is using `expiry_date`, so this is more consistent. :)

I guess if you wanted to be *really* consistent, you'd have to find an entirely new name for the field. 😜 One unclaimed phrase is `kick_the_cache_bucket_at`.

File: CRM/Core/PrevNextCache/Sql.php [hunk 4]
```diff
@@ -272,4 +293,10 @@ public function fetch($cacheKey, $offset, $rowCount) {
     return $cids;
   }
 
+  public function flush() {
+    // first find all the cacheKeys that match this
+    $sql = "DELETE FROM civicrm_prenext_cache WHERE expiry_date < NOW()";
```
Comment (totten) [inline CRM/Core/PrevNextCache/Sql.php]: v

**Comment by totten** [comment]
> (*@eileenmcnaughton wrote*) But when I looked at that I had doubts - we are going from setting the expiry at the key level to at the row (item within the key level) - with the granularity of a timestamp field we need to be really careful the rows don't save with slightly different expiry dates (probably you have dealt with that but it feels like a risk if we switch to tracking expiry by row).

Yeah, this is my high-level reservation too.

Just to verbalize the failure-concern: you fill the cache with a really nasty query, and it takes 30 seconds. The expiration times are spread out over a range of 30sec. There's a race if the user is still browsing the results around the time of expiration.

Maybe one could deal with that using a little more SQL in the cleanup function (pseudocode)?

```sql
CREATE TEMPORARY TABLE {$tmptable} AS
   SELECT DISTINCT cachekey FROM civicrm_prevnext_cache
   WHERE expiry_date < now();

DELETE FROM civicrm_prevnext_cache 
WHERE cachekey IN (SELECT cachekey FROM {$tmptable});
```

But I'm not sure about the performance characteristics... and not sure if there's a good way to adapt that to regular reads (i.e. IIRC, cleanups are async - so regular-reads need to be a bit defensive about skipping expired datasets).

**Comment by seamuslee001** [comment]
@totten in that case how does Redis work on a Cache Key level when it seems like you have 3 different cacheKeys working for the same issue here. I would have thought that given that whenver setItem is called that only one cacheKey is used, the other option is you could put an UPDATE set xxx where CACHE_KEY = x instead in a way right that way all rows with the same cache key have the same ttl in effect

**Comment by totten** [comment]
> in that case how does Redis work on a Cache Key level when it seems like you have 3 different cacheKeys working for the same issue here. 

Ah, yeah. The data model for `CRM_Core_PrevNextCache_Redis` is a bit different -- it's not just a basic key-value store. It might be better to think of it as 3 cached objects. Each object has a timeout, and each object can store multiple pieces of data. (There are different types of objects -- like basic strings; but also unsorted-sets, sorted-sets, hash-maps, etc).

Consider `CRM_Core_PrevNextCache_Redis::fillWithArray()`:

```php
    list($allKey, $dataKey, $selKey, $maxScore) = $this->initCacheKey($cacheKey);

    foreach ($rows as $row) {
      $maxScore++;
      $this->redis->zAdd($allKey, $maxScore, $row['entity_id1']);
      $this->redis->hSet($dataKey, $row['entity_id1'], $row['data']);
    }
   // ...
```

The call to `initCacheKey()` ensures that the three objects/keys exist. (It's always 3 objects/keys, regardless of the size the dataset.) It initializes (or updates) the expiration time for all 3 objects in rapid-succession (sub-second differences).  After the object is initialized/updated, then it does the `O(n)` work of inserting all the specific bits of data (e.g. the list of `$row['entity_id1']`s).

Which means that:

* The `$allKey`, `$dataKey`, and `$selKey` will expire at the same time (give-or-take sub-second variations in TTL issuance).
* When `$allKey` expires, it'll drop the entire list of `$row['entity_id1']`s stored under it.
* When `$dataKey` expires, it'll drop the entire list of `$row['data']`s stored under it.

File: CRM/Core/Payment/PayPalIPN.php (hunk 1 above)
Comment (mattwire) [reply CRM/Core/Payment/PayPalIPN.php]: @seamuslee001 Agree with @totten here - let's break this into a separate PR as it's a good change but has nothing to do with the expiry_date - and put the `convertDateToLocalTime()` function there too - we could get that merged quickly and then finish reviewing this one.

**Comment by mattwire** [comment]
@seamuslee001 A couple of general comments here:
1. I don't understand why we are doing so much timezone conversion for expiry_date - as it's a cache date surely we don't need to do any localtime conversion and could just use unixtime - the user will never see it.
2. Why are we doing the `(float)` microtime(1) stuff?  Is that to try and ensure some uniqueness? Bear in mind that casting to float means we'll always have tiny differences if comparing as floats are not completely precise in php.
3. As @totten says, we need to be really careful with race conditions.  The contact ACL cache suffers really badly with multiple attempts to update the cache at the same time.

File: CRM/Core/BAO/PrevNextCache.php [hunk 5]
```diff
@@ -200,8 +206,12 @@ public static function markConflict($id1, $id2, $cacheKey, $conflicts) {
         $pncUp = new CRM_Core_DAO_PrevNextCache();
         $pncUp->id = $pncFind->id;
         if ($pncUp->find(TRUE)) {
-          $pncUp->data     = serialize($data);
+          $expiry = (float) round(microtime(1)) + (float) self::DEFAULT_TTL;
+          $date = new DateTime();
+          $date->setTimeStamp($expiry);
+          $pncUp->data = serialize($data);
           $pncUp->cachekey = "{$cacheKey}_conflicts";
+          $pncUp->expiry_date = CRM_Utils_Date::convertDateToLocalTime($date, 'Y-m-d H:i:s');
```
Comment (eileenmcnaughton) [inline CRM/Core/BAO/PrevNextCache.php]: @mattwire is right - it doesn't make sense to convert dates when writing them to a timestamp field.

File: CRM/Core/DAO/PrevNextCache.php [hunk 6]
```diff
@@ -173,6 +180,19 @@ public static function &fields() {
...
           'localizable' => 0,
         ],
+        'expiry_date' => [
+          'name' => 'expiry_date',
+          'type' => CRM_Utils_Type::T_TIMESTAMP,
+          'title' => ts('Prev Next Expiry Date'),
+          'description' => ts('The Time this prev_next_cache data set expires'),
+          'required' => TRUE,
+          'where' => 'civicrm_prevnext_cache.expiry_date',
+          'default' => 'current_timestamp()',
```
Comment (eileenmcnaughton) [inline CRM/Core/DAO/PrevNextCache.php]: I think default should be NULL

File: CRM/Core/PrevNextCache/Sql.php [hunk 7]
```diff
@@ -97,37 +116,39 @@ public function markSelection($cacheKey, $action, $ids = NULL) {
...
       if ($action == 'select') {
         $whereClause .= "AND is_selected = 0";
-        $sql = "UPDATE civicrm_prevnext_cache SET is_selected = 1 {$whereClause}";
+        $sql = "UPDATE civicrm_prevnext_cache SET is_selected = 1, expiry_date = FROM_UNIXTIME(%3) {$whereClause}";
         $params[1] = [$cacheKey, 'String'];
       }
       elseif ($action == 'unselect') {
         $whereClause .= "AND is_selected = 1";
-        $sql = "UPDATE civicrm_prevnext_cache SET is_selected = 0 {$whereClause}";
+        $sql = "UPDATE civicrm_prevnext_cache SET is_selected = 0, expiry_date = FROM_UNIXTIME(%3) {$whereClause}";
```
Comment (eileenmcnaughton) [inline CRM/Core/PrevNextCache/Sql.php]: We shouldn't update the expiry date for one row within an 'expiry set'

File: CRM/Core/Payment/PayPalIPN.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Core/Payment/PayPalIPN.php]: In fact I would argue it it an unrelated change since we don't convert dates when writing to a timestamp field - mysql handles that

File: CRM/Core/BAO/PrevNextCache.php (hunk 5 above)
Comment (eileenmcnaughton) [reply CRM/Core/BAO/PrevNextCache.php]: Also - this is really grey. There are 2 uses for prev_next_cache - dedupe & search. I understood that they were currently treated somewhat differently - ie. only search uses the PrevNext interface. I think the dedupe caches persist for quite a long time - it feels like another can of worms

**Comment by seamuslee001** [comment]
@mattwire 

1. for the DAO->expiry_date field it wasn't accepting something like "FROM_UNIXTIME()" so had to go with a date time format and given that the TZ will be the user TZ and unixtime by default is in GMT have to convert to local time so that when it is stored into the DB itis the correct GMT. I.e with Syd AUs (+10H) if i just converted say a unix time stamp of `1562177433` it would convert to date Wednesday, July 3, 2019 6:10:33 PM and sending that to the date would cause the stored GMT date to be Wednesday, July 3, 2019 10:10:33 AM not Wednesday, July 3, 2019 6:10:33 PM

2. When testing this locally through unit tests it was complaining about non numeric but maybe i had code initially wrong can look at taking the floats out.

3. Agreed

File: CRM/Core/DAO/PrevNextCache.php (hunk 6 above)
Comment (seamuslee001) [reply CRM/Core/DAO/PrevNextCache.php]: which then means its a not required field i guess?

File: CRM/Core/BAO/PrevNextCache.php (hunk 5 above)
Comment (seamuslee001) [reply CRM/Core/BAO/PrevNextCache.php]: Does Dedupe call Civi::cache('prevnext') at all in SQL mode? I feel like dedupe should expire like Search does to me but agree that might be a meaninful change but would like @totten to discuss

File: CRM/Core/PrevNextCache/Sql.php (hunk 7 above)
Comment (seamuslee001) [reply CRM/Core/PrevNextCache/Sql.php]: This seems to be the case for the select key in redis as per https://github.com/civicrm/civicrm-core/blob/master/CRM/Core/PrevNextCache/Redis.php#L111 @totten

Comment (totten) [reply CRM/Core/PrevNextCache/Sql.php]: Sort of :)

* Consider a use-case for this bit of code: "*A user is browsing the result-set, checks some items, and then decides to "Clear all" selections. He then continues to select some other items.*" This operation is the "clear all selections".
* In the Redis.php example, it does reset the TTL, but that wasn't the main intent/goal. It was just side-effect that was hard to avoid.
    * The intent is to *clear out the cache-object which contains the full set of selected IDs* (`redis->delete($selKey)`). IIRC, this basically meant deleting it and re-creating it. To re-create the cache-object, one has to call `setTimeout()`.
    * It was a bit lazy/dirty on my part to use `self:TTL` for the re-created object b/c it effectively extends the timeout (beyond its original value). I rationalized that it's OK: the re-created `$selKey` will expire *after* the existing `$allKey`. We could save some memory by cleaning `$selKey` sooner, but we don't lose any functionality or data-integrity by keeping `$selKey` a bit longer. Why not? `$allKey` is the main driver of the screen. When `$allKey` expires, everything else is irrelevant.
* In the Sql.php example above, it extends the life of all data in the prevnext cache.
    * IMHO, it's OK either to keep the TTL (like it did before) or to renew it (like it does above). Renewing the TTL based on user-interaction basically gives the cache some LRU-ish characteristics. From a policy perspective, I think basic TTL or LRU or some arbitrary in-between thing will do the job. We tend to use TTL as a matter of KISS.

**Comment by mattwire** [comment]
@seamuslee001 Needs rebase

**Comment by eileenmcnaughton** [comment]
I'm also unclear if this is agreed in principle by @totten  - so whether to close or rebase is unclear to me at the moment

**Comment by eileenmcnaughton** [comment]
@seamuslee001 this is  stale -  I'm gonna close but also - it got stuck on me  being unclear  if it was  concept approved  so let's resolve   that first
